CUMULUS-3070 -- Refactor `granules.published` logic to allow for patch behavior

Jkovarik opened the pull request:
**Summary:** Summary of changes:

This is the first PR for CUMULUS-3070 - it updates the logic to allow 'publish' to be 'patched' in the API update call while still providing a default for all other cases. 


- **CUMULUS-3070**
  - Remove granules dynamoDb model logic that sets default publish value on record validation
  - Update API granule write logic to not set default publish value on record updates to avoid overwrite (PATCH behavior)
  - Update API granule write logic to  publish to false on record creation if not specified
  - Update message granule write logic to set default publish value on record creation update.
  - Removed dataType/version from api granule schema

Addresses [CUMULUS-3070: Develop amazing new feature](https://bugs.earthdata.nasa.gov/browse/CUMULUS-3070)

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Ad-hoc testing - Deploy changes and test manually
- [ ] Integration tests

Jkovarik commented on packages/api/models/schemas.js:
```diff
@@ -367,8 +366,6 @@ module.exports.granule = {
     timestamp: { type: 'integer' },
     createdAt: { type: 'integer' },
     updatedAt: { type: 'integer' },
-    dataType: { type: 'string' },
```
Consider removing this given it might result in a workflow failure

npauzenga commented on packages/api/endpoints/granules.js:
```diff
@@ -189,6 +192,9 @@ const putGranule = async (req, res) => {
   try {
+    if (isNewRecord === true && !apiGranule.published) {
+      apiGranule.published = false;
+    }
```
Thinking out loud... we default `published = false` whenever we're creating a granule (with POST or update) via the API endpoints. 

We do NOT default to `published = false` if we're creating a granule via workflow message or if we're migrating using the upsert logic in `upsertGranuleWithExecutionJoinRecord`.

I'm thinking we have it defaulting on the API endpoints but not the other avenues for creating Granules because it's not possible that we'll have an undefined `published`? But we can't guarantee that so I'm curious what you're thinking is here. Not saying anything is wrong, just trying to understand.

npauzenga replied:
Ok so you are defaulting `published = false` in the message write logic https://github.com/nasa/cumulus/pull/3112/files#diff-705c5c617b45462fde4dd4c0b8b964f95e91ec857b774841daebe5144d9dd990R873-R876. Cool. 

What about migrations? I think `published` can be null/undefined in Dynamo so do we want to default there too? Do we already and I just don't see it?

Jkovarik replied:
Discussed in huddle, agreed migration isn't impacted here and other considerations are out scope.

npauzenga commented on packages/api/lib/writeRecords/write-granules.js:
```diff
@@ -869,8 +870,13 @@ const writeGranulesFromMessage = async ({
       const updatedAt = now;
       const timestamp = now;
 
+      let published = granule.published;
+      if (published === undefined) {
+        published = false;
+      }
```
Do we need to handle `null` differently here like we do with PATCHing files? Or handle it at all I suppose. If it's null or undefined, published = false? I think you can have a `null` published in Dynamo?

Jkovarik replied:
Make sure unit covers null

Jkovarik replied:
@npauzenga  let me know what you think: c2c20b6f0d

npauzenga commented on packages/api/models/granules.js:
```diff
@@ -353,10 +353,11 @@ class Granule extends Manager {
    * @returns {Promise}
    */
   async _validateAndStoreGranuleRecord(granuleRecord) {
+    const clonedGranuleRecord = cloneDeep(granuleRecord);
```
Could we add a note here about why we're cloning it? Or change `recordIsValid` to not mutate (that seems way out of scope).

Jkovarik replied:
A note seems fine.    Refactoring to not use ajv .... is most definitely out of scope.

Jkovarik replied:
893273ca70

npauzenga replied:
Thanks. This is sort of unexpected behavior so I think it's good to have a note.

npauzenga commented on packages/api/models/schemas.js:
```diff
@@ -295,7 +295,6 @@ module.exports.granule = {
-      default: false,
       description: 'shows whether the granule is published to CMR',
       readonly: true,
     },
```
So I guess there aren't problems here if `published: undefined` even though that violates the type constraint. Is that because we'd need to call `recordIsValid` without `published` set to true/false and we've corrected the instances where that happens?

Jkovarik replied:
> So I guess there aren't problems here if published: undefined even though that violates the type constraint. 

off-hand, I believe in json boolean can be true/false/null.   Undefined should be parsed to 'missing key'. 

A test in the granule validation units, as an example, passes according to the comments below: 


```
test.only('storeGranule() fails validation if granule published is <value>', async (t) => {
  const {
    granuleModel,
    collectionId,
    provider,
    workflowStatus,
  } = t.context;

  const bucket = randomId('bucket');
  await S3.createBucket(bucket);
  t.teardown(() => S3.recursivelyDeleteS3Bucket(bucket));

  const granule1 = fakeGranuleFactoryV2({
    files: [fakeFileFactory({ bucket })],
  });

  await S3.s3PutObject({ Bucket: bucket, Key: granule1.files[0].key, Body: 'asdf' });

  const files = await granuleModel.fileUtils.buildDatabaseFiles({
    s3: awsClients.s3(),
    providerURL: buildURL(provider),
    files: granule1.files,
  });

  const granuleRecord = await generateGranuleApiRecord({
    granule: granule1,
    files,
    executionUrl: 'http://execution-url.com',
    collectionId,
    provider: provider.id,
    workflowStatus,
    status: getGranuleStatus(workflowStatus, granule1),
    cmrUtils,
    updatedAt: Date.now(),
  });

  // delete granuleRecord.published; //-- pass
  // granuleRecord.published = false; -- pass
  // granuleRecord.published = undefined; -- pass
  await granuleModel.storeGranule(granuleRecord);

  t.true(await granuleModel.exists({ granuleId: granule1.granuleId }));
});
```

Jkovarik replied:
Note - ^ was off, null fails.

Jkovarik replied:
Updated above to address the nullish delete to false concern.

npauzenga commented on packages/api/tests/endpoints/test-granules.js:
```diff
@@ -1976,6 +1978,113 @@ test.serial('PUT replaces an existing granule in all data stores', async (t) =>
+  });
+
+  t.deepEqual(
+    omitBy(actualPgGranule, (x) => x === null),
```
Worth having a helper that only removes `nulls` instead of null + undefined?

npauzenga replied:
Also maybe we should use an explicit `omitList` here?

Jkovarik replied:
I think in this instance it's just a function of using knex.get that undefineds come back as null.    Given that, I'm opting to go with: bb20d73e5b .  Let me know if that works.

npauzenga commented on packages/api/tests/endpoints/test-granules.js:
```diff
@@ -1976,6 +1978,113 @@ test.serial('PUT replaces an existing granule in all data stores', async (t) =>
+
+  t.is(actualPgGranule.updated_at.getTime(), actualGranule.updatedAt);
+
+  const updatedEsRecord = await t.context.esGranulesClient.get(
```
I think this is a new record, not `updated`, right?

Jkovarik replied:
79ac1bd9ae

npauzenga commented on packages/api/tests/endpoints/test-granules.js:
```diff
@@ -1976,6 +1978,113 @@ test.serial('PUT replaces an existing granule in all data stores', async (t) =>
+  );
+});
+
+test.serial('PUT sets correct default value for published on new granule creation', async (t) => {
```
```suggestion
test.serial('PUT sets correct default value for published on new granule creation if no value is provided', async (t) => {
```

npauzenga commented on packages/api/tests/lib/writeRecords/test-write-granules.js:
```diff
@@ -2048,6 +2100,62 @@ test.serial('writeGranuleFromApi() given a partial granule overwrites only provi
+  t.deepEqual(
+    omit(removeNilProperties(pgGranule), apiOmitList),
+    omit(removeNilProperties({ ...originalpgGranule, ...updatedPgGranuleFields }), apiOmitList)
+  );
```
This catches it but what if we scoped this test down a bit and just checked that `pgGranule.published === originalPgGranule.published`?

Jkovarik replied:
fcc71b620d

npauzenga replied:
Thanks I think it's a little clearer but what you had previously did the job too 🤷

npauzenga commented on packages/message/src/Granules.ts:
```diff
@@ -246,15 +246,15 @@ export const generateGranuleApiRecord = async ({
     granuleId,
     cmrLink,
-    published = false,
+    published,
```
So the idea here is to set the defaults as far up as possible, yeah? I think that's a good call having this logic in the endpoint and message write logic.

It makes me a little nervous that we have `writeGranuleFromApi` and it's callers `createGranuleFromApi` + `updateGranuleFromApi` available as exported functions that don't set the defaults correctly but... that may be OK. They don't appear to be used in a way that would cause problems now, only in the endpoints you've updated. In the future if someone were to call e.g. `writeGranuleFromApi` directly without a `published` field, there might be problems so we'd have to know not to do that.

Jkovarik replied:
> It makes me a little nervous that we have writeGranuleFromApi and it's callers createGranuleFromApi + updateGranuleFromApi available as exported functions that don't set the defaults correctly but... that may be OK.

That's not entirely unfair - if we feel that those methods should defend against the defaults being set in non patch situations we could write a wrapper that throws (?).  

That said, I think we want the write logic chain  to know as little about business-logic defaults as possible, particularly in P3, as the database *schema* should be handling fall-through validation concerns, and I think we should be calling what's in the api model for API record validation before getting to the db write phase.  

Meaning the business of setting defaults/incoming validation should be in the input/output endpoint logic and/or message write logic, with the write then being forced to conform to the schema in the postgres DB by virtue of the write, *generally*.        It's possible to also set defaults within the db schema, although that has issues with the current logic and replication to dynamo, and is inconsistent with the approach.

Jkovarik replied:
Agreed to punt to an arch meeting/P3.

Jkovarik commented:
@npauzenga  thanks for the review, and your patience with the updates.  I think I've addressed your feedback, take a look and let me know what you think!

botanical commented on packages/api/lib/writeRecords/write-granules.js:
```diff
@@ -723,7 +724,13 @@ const writeGranuleFromApi = async (
+    // instead of allowing nullish value
+    let publishedValue = published;
+    if (published === null) {
+      publishedValue = false;
```
👍

botanical commented on packages/api/lib/writeRecords/write-granules.js:
```diff
@@ -869,8 +876,13 @@ const writeGranulesFromMessage = async ({
 
+      let published = granule.published;
+      if (isNil(published)) {
+        published = false;
```
Just out of curiosity, is there a reason that `writeGranuleFromApi`'s published check doesn't use `isNil`?

npauzenga reviewed (approved):
Thanks @Jkovarik you hit all of my comments. @botanical feel free to do an independent review. This stuff is complicated so more eyes are better.

botanical commented on packages/api/tests/endpoints/test-granules.js:
```diff
@@ -1976,6 +1979,113 @@ test.serial('PUT replaces an existing granule in all data stores', async (t) =>
+    granuleId,
+    status: 'completed',
+    execution: t.context.executionUrl,
+    duration: 47.125,
```
the collection created from the `fakeGranuleFactoryV2` record happens to have the same collection name and version as the collection specific in the test setup 
```
  const collectionName = 'fakeCollection';
  const collectionVersion = 'v1';
```

but if these were somehow to change in the fakeGranuleFactoryV2 function, than the call here 
```
  const actualPgGranule = await t.context.granulePgModel.get(knex, {
    granule_id: actualGranule.granuleId,
    collection_cumulus_id: t.context.collectionCumulusId,
  });
```
could be getting the wrong granule, right? Not a big deal but I think we could explicitly pass in the collection name and version here instead of relying on the values defined in `fakeGranuleFactoryV2`

botanical commented on packages/api/tests/lib/writeRecords/test-write-granules.js:
```diff
@@ -2907,3 +3010,91 @@ test.serial('updateGranuleStatusToFailed() throws error if record does not exist
   );
 });
+
+test.serial('writeGranuleFromApi() saves granule record with publish set to null with publish value set to false to all datastores', async (t) => {
```
```suggestion
test.serial('writeGranuleFromApi() saves granule record and sets `published` to false if null value is set', async (t) => {
```
This is a nit but I had a hard time understanding the test statement 😅

botanical commented on packages/api/tests/lib/writeRecords/test-write-granules.js:
```diff
@@ -2907,3 +3010,91 @@ test.serial('updateGranuleStatusToFailed() throws error if record does not exist
+  t.deepEqual(omit(translatedPgRecord, omitList), omit(esRecord, omitList));
+
+  // Validate assertion is true in the primary datastore:
+
```
Nit, newline unnecessary
```suggestion
```

botanical commented on packages/api/endpoints/granules.js:
```diff
@@ -130,6 +130,9 @@ const create = async (req, res) => {
     return res.boom.badRequest(errorify(error));
   } try {
+    if (!granule.published) {
+      granule.published = false;
```
👍
